pkg/util/objectwatcher/objectwatcher.go

```diff
@@ -254,20 +262,20 @@ func (o *objectWatcherImpl) deleteVersionRecord(clusterName, resourceName string
 	delete(o.VersionRecord[clusterName], resourceName)
 }
 
-func (o *objectWatcherImpl) NeedsUpdate(clusterName string, desiredObj, clusterObj *unstructured.Unstructured) (bool, error) {
-	// get resource version
+func (o *objectWatcherImpl) NeedsUpdate(clusterName string, desiredObj, clusterObj *unstructured.Unstructured) bool {
```

Comment:
Perhaps we don't need this method.

Reply:
As https://github.com/karmada-io/karmada/pull/3808#discussion_r1274514729, I still think comparing the `generation` between current object and recorded since last updating is a good and explicit way to tell us if the object has been modified unexpectly.

Reply:
Yes, I agree with you. Sorry, I didn't make it clear. My point is that we can remove the version recorded in memory and directly compare the generations of the two objects.

Reply:
Yes, you are right. I always imagine too much about some special cases not in a normal workflow. Maybe we could talk about these cases in a new issue. Here I will do some simple comparation. I still expect for #3804! :)

Reply:
Sorry, I just came up with that an important reason why I attempted to reuse this part:   

Comparing the `generation` in memory could prevent requeueing from the update event triggered by `karmada` self.

If we overlook this situation, we may have two attempts to update the object in a short time after once modification.   

What do you think about this situation?

Reply:
I'm sorry, I didn't understand that.

Reply:
For example, we have an object managed by `karmada` with `generation = 1` now in member cluster.  

Then I unexpectly modified the `spec` field of the object in member cluster. Now it is with `generation = 2`. This would trigger the updating from `execution controller` because of the changing of `generation`.  

The `execution controller` would sync the object in member cluster. Now the object is with `generation = 3`. This would immediately trigger another updating from `execution controller` again, and it is still because of the changing of `generation`.  

Besides, even the first updating is from the modification of `work`, the `execution controller` would attempt to update the  object twice because it would watch the changing of `generation` in object in member cluster caused by the modification by itself.

Reply:
Thanks, I got it.

Reply:
Do you have any advices about this? Or we had better keep use `ObjectWatcher.NeedsUpdate`?

Reply:
I think it can still be used, but the parameter names need to be modified in order to make them more consistent with their current meaning.

Reply:
Done.  

And I also made some modifications on the way to record version, to prevent enqueueing object on update event too early when the updated version hasn't been recorded yet.